File: src/main/java/com/netsensia/rivalchess/engine/core/hash/BoardHash.java
package com.netsensia.rivalchess.engine.core.hash;

import com.netsensia.rivalchess.bitboards.Bitboards;
import com.netsensia.rivalchess.config.Evaluation;
import com.netsensia.rivalchess.config.FeatureFlag;
import com.netsensia.rivalchess.config.Hash;
import com.netsensia.rivalchess.model.Colour;
import com.netsensia.rivalchess.engine.core.EngineChessBoard;
import com.netsensia.rivalchess.engine.core.type.EngineMove;
import com.netsensia.rivalchess.enums.HashIndex;
import com.netsensia.rivalchess.enums.HashValueType;
import com.netsensia.rivalchess.enums.PawnHashIndex;
import com.netsensia.rivalchess.model.SquareOccupant;

public class BoardHash {

    private final ZorbristHashTracker hashTracker = new ZorbristHashTracker();

    private int hashTableVersion;
    private int[] hashTableUseHeight;
    private int[] hashTableIgnoreHeight;
    private long[] pawnHashTable;
    private int maxHashEntries;
    private int lastHashSizeCreated;

    public int getHashTableUseHeight(int index) {
        return hashTableUseHeight[index];
    }

    public int getHashTableIgnoreHeight(int index) {
        return hashTableIgnoreHeight[index];
    }

    public void setHashTableUseHeight(int index, int value) {
        hashTableUseHeight[index] = value;
    }

    public void setPawnHashTable(int index, long value) {
        pawnHashTable[index] = value;
    }

    public synchronized void clearHash() {
        for (int i = 0; i < maxHashEntries; i++) {
            this.hashTableUseHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_FLAG.getIndex()] = HashValueType.EMPTY.getIndex();
            this.hashTableUseHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_HEIGHT.getIndex()] = Hash.DEFAULT_SEARCH_HASH_HEIGHT.getValue();
            this.hashTableIgnoreHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_FLAG.getIndex()] = HashValueType.EMPTY.getIndex();
            this.hashTableIgnoreHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_HEIGHT.getIndex()] = Hash.DEFAULT_SEARCH_HASH_HEIGHT.getValue();
            if (FeatureFlag.USE_PAWN_HASH.isActive()) {
                this.pawnHashTable[i * PawnHashIndex.getNumHashFields() + PawnHashIndex.PAWNHASHENTRY_MAIN_SCORE.getIndex()] = -Integer.MAX_VALUE;
            }
        }
    }

    public synchronized void setHashTable() {
        if (maxHashEntries != lastHashSizeCreated) {
            this.hashTableUseHeight = new int[maxHashEntries * HashIndex.getNumHashFields()];
            this.hashTableIgnoreHeight = new int[maxHashEntries * HashIndex.getNumHashFields()];
            if (FeatureFlag.USE_PAWN_HASH.isActive()) {
                this.pawnHashTable = new long[maxHashEntries * PawnHashIndex.getNumHashFields()];
            }
            lastHashSizeCreated = maxHashEntries;
            for (int i = 0; i < maxHashEntries; i++) {
                this.hashTableUseHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_FLAG.getIndex()] = HashValueType.EMPTY.getIndex();
                this.hashTableUseHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_HEIGHT.getIndex()] = Hash.DEFAULT_SEARCH_HASH_HEIGHT.getValue();
                this.hashTableUseHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_VERSION.getIndex()] = 1;
                this.hashTableIgnoreHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_FLAG.getIndex()] = HashValueType.EMPTY.getIndex();
                this.hashTableIgnoreHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_HEIGHT.getIndex()] = Hash.DEFAULT_SEARCH_HASH_HEIGHT.getValue();
                this.hashTableIgnoreHeight[i * HashIndex.getNumHashFields() + HashIndex.HASHENTRY_VERSION.getIndex()] = 1;
                if (FeatureFlag.USE_PAWN_HASH.isActive()) {
                    this.pawnHashTable[i * PawnHashIndex.getNumHashFields() + PawnHashIndex.PAWNHASHENTRY_MAIN_SCORE.getIndex()] = -Integer.MAX_VALUE;
                }
            }
        }
    }

    public void storeHashMove(int move, EngineChessBoard board, int score, byte flag, int height) {
        final int hashIndex = (int) (board.trackedBoardHashCode() % maxHashEntries) * HashIndex.getNumHashFields();

        if (height >= this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.getIndex()] || hashTableVersion > this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()]) {
            if (hashTableVersion == this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()]) {
                copyEntryFromUseHeightToIgnoreHeightTable(hashIndex);
            }

            storeSuperVerifyUseHeightInformation(board, hashIndex);

            storeMoveInHashTable(move, board, score, flag, height, hashIndex, this.hashTableUseHeight);
        } else {
            storeSuperVerifyAlwaysReplaceInformation(board, hashIndex);

            storeMoveInHashTable(move, board, score, flag, height, hashIndex, this.hashTableIgnoreHeight);
        }
    }

    private void copyEntryFromUseHeightToIgnoreHeightTable(int hashIndex) {
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_MOVE.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_MOVE.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_SCORE.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_SCORE.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_FLAG.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_FLAG.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.getIndex()];
        this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()];
    }

    private void storeMoveInHashTable(int move, EngineChessBoard board, int score, byte flag, int height, int hashIndex, int[] hashTableUseHeight) {
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_MOVE.getIndex()] = move;
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_SCORE.getIndex()] = score;
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_FLAG.getIndex()] = flag;
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.getIndex()] = (int) (board.trackedBoardHashCode() >>> 32);
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.getIndex()] = (int) (board.trackedBoardHashCode() & Bitboards.LOW32);
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.getIndex()] = height;
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()] = hashTableVersion;
    }

    private void storeSuperVerifyAlwaysReplaceInformation(EngineChessBoard board, final int hashIndex) {
        if (FeatureFlag.USE_SUPER_VERIFY_ON_HASH.isActive()) {
            for (int i = SquareOccupant.WP.getIndex(); i <= SquareOccupant.BR.getIndex(); i++) {
                this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i] = (int) (board.getBitboardByIndex(i) >>> 32);
                this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i + 12] = (int) (board.getBitboardByIndex(i) & Bitboards.LOW32);
            }
        }
    }

    private void storeSuperVerifyUseHeightInformation(EngineChessBoard board, final int hashIndex) {
        if (FeatureFlag.USE_SUPER_VERIFY_ON_HASH.isActive()) {
            for (int i = SquareOccupant.WP.getIndex(); i <= SquareOccupant.BR.getIndex(); i++) {
                if (hashTableVersion == this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_VERSION.getIndex()]) {
                    this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i];
                    this.hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i + 12] = this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i + 12];
                }
                this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i] = (int) (board.getBitboardByIndex(i) >>> 32);
                this.hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.getIndex() + i + 12] = (int) (board.getBitboardByIndex(i) & Bitboards.LOW32);
            }
        }
    }

    public synchronized void setHashSizeMB(int hashSizeMB) {
        if (hashSizeMB < 1) {
            setMaxHashEntries(1);
        } else {
            int mainHashTableSize = ((hashSizeMB * 1024 * 1024) / 14) * 6; // two of these
            setMaxHashEntries(mainHashTableSize / HashIndex.getHashPositionSizeBytes());
        }

        setHashTable();
    }

    public synchronized void initialiseHashCode(EngineChessBoard engineChessBoard) {
        hashTracker.initHash(engineChessBoard);
    }

    public int getHashIndex(EngineChessBoard engineChessBoard) {
        return getHashIndex(engineChessBoard.trackedBoardHashCode());
    }

    public int getHashIndex(long hashValue) {
        return (int) (hashValue % maxHashEntries) * HashIndex.getNumHashFields();
    }

    public void move(EngineChessBoard engineChessBoard, EngineMove move) {
        hashTracker.makeMove(engineChessBoard, move);
    }

    public void unMove(EngineChessBoard engineChessBoard) {
        hashTracker.unMakeMove(engineChessBoard.getLastMoveMade());
    }

    public void makeNullMove() {
        hashTracker.nullMove();
    }

    public long getTrackedHashValue() {
        return hashTracker.getTrackedBoardHashValue();
    }

    public long getTrackedPawnHashValue() {
        return hashTracker.getTrackedPawnHashValue();
    }

    public int getHashTableVersion() {
        return hashTableVersion;
    }

    public void setHashTableVersion(int hashTableVersion) {
        this.hashTableVersion = hashTableVersion;
    }

    public void setMaxHashEntries(int maxHashEntries) {
        this.maxHashEntries = maxHashEntries;
    }

    public void incVersion() {
        this.hashTableVersion++;
    }
}

File: src/main/java/com/netsensia/rivalchess/engine/core/hash/BoardHash.kt
package com.netsensia.rivalchess.engine.core.hash

import com.netsensia.rivalchess.bitboards.LOW32
import com.netsensia.rivalchess.config.DEFAULT_SEARCH_HASH_HEIGHT
import com.netsensia.rivalchess.config.USE_SUPER_VERIFY_ON_HASH
import com.netsensia.rivalchess.engine.core.board.EngineBoard
import com.netsensia.rivalchess.engine.core.type.EngineMove
import com.netsensia.rivalchess.enums.HashIndex
import com.netsensia.rivalchess.enums.HashValueType
import com.netsensia.rivalchess.model.SquareOccupant

class BoardHash {

    private val hashTracker = ZobristHashTracker()

    var hashTableVersion = 0

    private lateinit var hashTableUseHeight: IntArray
    private lateinit var hashTableIgnoreHeight: IntArray
    private var maxHashEntries = 0
    private var lastHashSizeCreated = 0

    fun useHeight(index: Int) = hashTableUseHeight[index]
    fun ignoreHeight(index: Int) = hashTableIgnoreHeight[index]

    fun setHashTableUseHeightVersion(index: Int, value: Int) {
        hashTableUseHeight[index + HashIndex.VERSION.index] = value
    }

    fun clearHash() {
        for (i in 0 until maxHashEntries) {
            hashTableUseHeight[i * HashIndex.numHashFields + HashIndex.FLAG.index] = HashValueType.EMPTY.index
            hashTableUseHeight[i * HashIndex.numHashFields + HashIndex.HASHENTRY_HEIGHT.index] = DEFAULT_SEARCH_HASH_HEIGHT
            hashTableIgnoreHeight[i * HashIndex.numHashFields + HashIndex.FLAG.index] = HashValueType.EMPTY.index
            hashTableIgnoreHeight[i * HashIndex.numHashFields + HashIndex.HASHENTRY_HEIGHT.index] = DEFAULT_SEARCH_HASH_HEIGHT
        }
    }

    fun setHashTable() {
        if (maxHashEntries != lastHashSizeCreated) {
            hashTableUseHeight = IntArray(maxHashEntries * HashIndex.numHashFields)
            hashTableIgnoreHeight = IntArray(maxHashEntries * HashIndex.numHashFields)
            lastHashSizeCreated = maxHashEntries
            for (i in 0 until maxHashEntries) {
                hashTableUseHeight[i * HashIndex.numHashFields + HashIndex.FLAG.index] = HashValueType.EMPTY.index
                hashTableUseHeight[i * HashIndex.numHashFields + HashIndex.HASHENTRY_HEIGHT.index] = DEFAULT_SEARCH_HASH_HEIGHT
                hashTableUseHeight[i * HashIndex.numHashFields + HashIndex.VERSION.index] = 1
                hashTableIgnoreHeight[i * HashIndex.numHashFields + HashIndex.FLAG.index] = HashValueType.EMPTY.index
                hashTableIgnoreHeight[i * HashIndex.numHashFields + HashIndex.HASHENTRY_HEIGHT.index] = DEFAULT_SEARCH_HASH_HEIGHT
                hashTableIgnoreHeight[i * HashIndex.numHashFields + HashIndex.VERSION.index] = 1
            }
        }
    }

    fun storeHashMove(move: Int, board: EngineBoard, score: Int, flag: Byte, height: Int) {
        val hashIndex = (board.trackedBoardHashCode() % maxHashEntries).toInt() * HashIndex.numHashFields
        if (height >= hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.index] || hashTableVersion > hashTableUseHeight[hashIndex + HashIndex.VERSION.index]) {
            if (hashTableVersion == hashTableUseHeight[hashIndex + HashIndex.VERSION.index]) {
                copyEntryFromUseHeightToIgnoreHeightTable(hashIndex)
            }
            storeSuperVerifyUseHeightInformation(board, hashIndex)
            storeMoveInHashTable(move, board, score, flag, height, hashIndex, hashTableUseHeight)
        } else {
            storeSuperVerifyAlwaysReplaceInformation(board, hashIndex)
            storeMoveInHashTable(move, board, score, flag, height, hashIndex, hashTableIgnoreHeight)
        }
    }

    private fun copyEntryFromUseHeightToIgnoreHeightTable(hashIndex: Int) {
        hashTableIgnoreHeight[hashIndex + HashIndex.MOVE.index] = hashTableUseHeight[hashIndex + HashIndex.MOVE.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.SCORE.index] = hashTableUseHeight[hashIndex + HashIndex.SCORE.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.FLAG.index] = hashTableUseHeight[hashIndex + HashIndex.FLAG.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.index] = hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.index] = hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.index] = hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.index]
        hashTableIgnoreHeight[hashIndex + HashIndex.VERSION.index] = hashTableUseHeight[hashIndex + HashIndex.VERSION.index]
    }

    private fun storeMoveInHashTable(move: Int, board: EngineBoard, score: Int, flag: Byte, height: Int, hashIndex: Int, hashTableUseHeight: IntArray) {
        hashTableUseHeight[hashIndex + HashIndex.MOVE.index] = move
        hashTableUseHeight[hashIndex + HashIndex.SCORE.index] = score
        hashTableUseHeight[hashIndex + HashIndex.FLAG.index] = flag.toInt()
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT1.index] = (board.trackedBoardHashCode() ushr 32).toInt()
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_64BIT2.index] = (board.trackedBoardHashCode() and LOW32).toInt()
        hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_HEIGHT.index] = height
        hashTableUseHeight[hashIndex + HashIndex.VERSION.index] = hashTableVersion
    }

    private fun storeSuperVerifyAlwaysReplaceInformation(board: EngineBoard, hashIndex: Int) {
        if (USE_SUPER_VERIFY_ON_HASH) {
            for (i in SquareOccupant.WP.index..SquareOccupant.BR.index) {
                hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i] = (board.getBitboard(i) ushr 32).toInt()
                hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i + 12] = (board.getBitboard(i) and LOW32).toInt()
            }
        }
    }

    private fun storeSuperVerifyUseHeightInformation(board: EngineBoard, hashIndex: Int) {
        if (USE_SUPER_VERIFY_ON_HASH) {
            for (i in SquareOccupant.WP.index..SquareOccupant.BR.index) {
                if (hashTableVersion == hashTableUseHeight[hashIndex + HashIndex.VERSION.index]) {
                    hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i] = hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i]
                    hashTableIgnoreHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i + 12] = hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i + 12]
                }
                hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i] = (board.getBitboard(i) ushr 32).toInt()
                hashTableUseHeight[hashIndex + HashIndex.HASHENTRY_LOCK1.index + i + 12] = (board.getBitboard(i) and LOW32).toInt()
            }
        }
    }

    fun setHashSizeMB(hashSizeMB: Int) {
        if (hashSizeMB < 1) {
            setMaxHashEntries(1)
        } else {
            val mainHashTableSize = hashSizeMB * 1024 * 1024 / 14 * 6 // two of these
            setMaxHashEntries(mainHashTableSize / HashIndex.hashPositionSizeBytes)
        }
        setHashTable()
    }

    fun initialiseHashCode(engineBoard: EngineBoard) {
        hashTracker.initHash(engineBoard)
    }

    fun getHashIndex(engineBoard: EngineBoard): Int {
        return getHashIndex(engineBoard.trackedBoardHashCode())
    }

    fun getHashIndex(hashValue: Long): Int {
        return (hashValue % maxHashEntries).toInt() * HashIndex.numHashFields
    }

    fun move(move: EngineMove, movePiece: SquareOccupant, capturePiece: SquareOccupant) {
        hashTracker.makeMove(move, movePiece, capturePiece)
    }

    fun unMove(engineBoard: EngineBoard) {
        hashTracker.unMakeMove(engineBoard.lastMoveMade)
    }

    fun makeNullMove() {
        hashTracker.nullMove()
    }

    val trackedHashValue: Long
        get() = hashTracker.trackedBoardHashValue

    private fun setMaxHashEntries(maxHashEntries: Int) {
        this.maxHashEntries = maxHashEntries
    }

    fun incVersion() {
        hashTableVersion++
    }
}